src/grt/src/Pin.cpp

```diff
@@ -95,10 +95,17 @@ void Pin::determineEdge(const odb::Rect& bounds,
                         const odb::Point& pin_position,
                         const std::vector<odb::dbTechLayer*>& layers)
 {
+  odb::Point upper_right = pin_position;
+  // for IO pins, use the position that touches the die edge to detect the edge
+  // it is placed
+  if (is_port_) {
+    const int layer = layers_.back();
+    upper_right = boxes_per_layer_[layer].back().ur();
+  }
```

Comment:
Could something similar happen with a macro pin?

Reply:
Yes, I've included macro pins in this check. My only concern is about the pin shapes of a macro pin. Can we have a macro pin with multiple pin shapes in different layers? Do we have something like this in our designs?

Reply:
You can have a multi-layer pin in LEF (iterm or bterm).  I don't know if we have any test cases of such

Reply:
I've updated the code to iterate over all the pin shapes, and choose the pin edge in a voting system. I believe it will be more accurate than just getting the top layer's shape.